src/transform/deadCode.ml

```diff
@@ -0,0 +1,217 @@
+    Option.is_some skind'_opt
+  in
+  impl_block block
+
+(** Perform a depth first search over the CFG. Record the IDs of live statements;
+    for each traversed edge, record the skipped statements along the edge as live,
+    if the nodes on both ends of the edge are live. Record live statements in the nodes
+    themselves as well. *)
+let find_live_statements
+    (node_live : node -> bool)
+    (skipped_statements : node -> edges -> node -> stmt list)
+    (cfg : cfg) (start : node)
+  =
+  let seen_nodes = NodeH.create 13 in
+  let module IS = BatSet.Int in
+  let node_id_set = function Statement stmt -> IS.singleton stmt.sid | _ -> IS.empty in
+
+  let rec impl (n : node) (ns : node list) (live_stmts : IS.t) =
+    NodeH.replace seen_nodes n ();
+    let n_outbound = cfg n in
+
+    (* The skipped statements along each edge could be live if either the 'from'
+      or the 'to' nodes are live. This is an over-approximation; if the 'from'
+      node is live but never completes normally (e.g. a call to abort), and the
+      'to' node is live from another path, the statements will not be marked as dead. *)
+    let live_stmts' =
+      (* check if current node is dead: otherwise all outgoing edges are dead anyway *)
+      if node_live n then
+        n_outbound
+        |> List.filter (node_live % snd)
+        |> List.map (fun (edges, n') ->
+            IS.union
+              (node_id_set n')
+              (skipped_statements n edges n' |> List.map (fun stmt -> stmt.sid) |> IS.of_list))
+        |> List.fold_left IS.union live_stmts
+      else
+        live_stmts
+    in
+
+    match (n_outbound |> List.map snd |> List.filter (not % NodeH.mem seen_nodes)) @ ns with
+    | [] -> live_stmts'
+    | n' :: ns' -> impl n' ns' live_stmts'
+  in
+
+  impl start [] (if node_live start then node_id_set start else IS.empty)
+
+(* TODO: does this not exist already? *)
+(** A fresh global name, of the form 'name' (if already unused), otherwise 'name_N'.
+    A call to [Cil.uniqueVarNames] may still be necessary to handle conflicting locals. *)
+let fresh_global_name file name =
+  let used_names = Hashtbl.create 7 in
+  Cil.iterGlobals file (function
+    | GVarDecl (vi, _) | GVar (vi, _, _) | GFun ({ svar = vi; _ }, _) ->
+      Hashtbl.replace used_names vi.vname ()
+    | GType (ti, _) ->
+      Hashtbl.replace used_names ti.tname ()
+    | _ -> () );
+  let with_suffix = function 0 -> name | n -> name ^ "_" ^ string_of_int n in
+  Seq.(ints 0 |> map with_suffix |> filter (not % Hashtbl.mem used_names) |> first)
+
+module RemoveDeadCode : Transform.S = struct
+  let transform (q : Transform.queries) (file : file) : unit =
+
+    (* whether a global function (might) still be live, and should therefore be kept *)
+    let fundec_live : fundec -> bool = not % q.must_be_uncalled in
+
+    (* add an uninitialized global variable to be used as the condition
+       in if-statements where the condition is never checked *)
```

Comment:
Why is this needed? If it's unreachable, then it can be anything, e.g. `1`.
Synthesizing a new global variable (and avoiding name conflicts) is a lot of extra work but that global is still default initialized to 0, so it's not any better than a direct constant condition.

Reply:
The idea was to document in the resulting code that the condition is dead; there's no functional benefit. But given the amount of work to synthesize the global, it might not be worth it. Really, a comment above the conditional would be enough, but there's no way to do that with CIL.

Reply:
I would just use the constant `1` condition and remove the extra logic for creating these globals, but maybe @michael-schwarz has a different opinion?

Reply:
I think it makes sense to have this special global to draw attention to this fact, that it is in fact never checked. That might make reduced code easier to understand as it is clear that the things inside can only be reachable by jumps etc.

Reply:
A simpler alternative would be to use a string literal like `"dead condition"`. It contains the same kind of description and is constant like the global variable.